## Dataquest.io Guided Project
# Exploring Ebay Car Sales Data
In this guided project, we'll work with a dataset of used cars from *ebay Kleinanzeigen*, a [classified]("https://en.wikipedia.org/wiki/Classified_advertisinghttps://en.wikipedia.org/wiki/Classified_advertising") section of the German ebay website.

The dataset was originally [scraped]("https://en.wikipedia.org/wiki/Web_scraping") and uploaded to [Kaggle]("https://www.kaggle.com/orgesleka/used-cars-database/data"). We've made a few modifications from the original dataset that was uploaded to Kaggle:

* We sampled 50,000 data points from the full dataset, to ensure your code runs quickly in our hosted environment

* We dirtied the dataset a bit to more closely resemble what you would expect from a scraped dataset (the version uploaded to Kaggle was cleaned to be easier to work with)


The data dictionary provided with data is as follows:

* **dateCrawled** - When this ad was first crawled. All field-values are taken from this date.
* **name** - Name of the car.
* **seller** - Whether the seller is private or a dealer.
* **offerType** - The type of listing
* **price** - The price on the ad to sell the car.
* **abtest** - Whether the listing is included in an A/B test.
* **vehicleType** - The vehicle Type.
* **yearOfRegistration** - The year in which the car was first registered.
* **gearbox** - The transmission type.
* **powerPS** - The power of the car in PS.
* **model** - The car model name.
* **kilometer** - How many kilometers the car has driven.
* **monthOfRegistration** - The month in which the car was first registered.
* **fuelType** - What type of fuel the car uses.
brand - The brand of the car.
* **notRepairedDamage** - If the car has a damage which is not yet repaired.
* **dateCreated** - The date on which the eBay listing was created.
* **nrOfPictures** - The number of pictures in the ad.
* **postalCode** - The postal code for the location of the vehicle.
* **lastSeenOnline** - When the crawler saw this ad last online.

The aim of this project is to clean the data and analyze the included used car listings. You'll also become familiar with some of the unique benefits jupyter notebook provides for pandas.





In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
autos = pd.read_csv('autos.csv', encoding='Latin-1')


In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [4]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

The data set is made of up 50000 rows and 20 columns. within the columns, two data types dominate: object(15) which is also known as string and integers(5). We can also see that the memory usage of the data set is high.

In [5]:
autos.columns 

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:

autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_PS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_of_pictures', 'postal_code',
       'last_seen']

In [7]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_PS   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00                0   
1         bmw              nein  2016-04-04 00:00:00                0   
2  volkswagen              nein  2016-03-26 00:00:00                0   
3       smart              nein  2016-03-12 00:00:00                0   
4        ford              nein  2016-04-01 00:00:00                0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

We adjusted the columns from camelcase to pythonic snakecase as this would ease typing and aids  referencei in the case of applying dot notation using the columns

In [8]:
autos.describe()

registration_year      power_PS  registration_month  num_of_pictures  \
count       50000.000000  50000.000000        50000.000000          50000.0   
mean         2005.073280    116.355920            5.723360              0.0   
std           105.712813    209.216627            3.711984              0.0   
min          1000.000000      0.000000            0.000000              0.0   
25%          1999.000000     70.000000            3.000000              0.0   
50%          2003.000000    105.000000            6.000000              0.0   
75%          2008.000000    150.000000            9.000000              0.0   
max          9999.000000  17700.000000           12.000000              0.0   

        postal_code  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

There are no pictures hence we would drop the number of picture column from the data set.

The price isn't seen under the descritive statistics;it need to be cleaned and turned to the required data type.

similarly offertype and seller columns need investigating and possibly removal

In [11]:
autos['num_of_pictures'].value_counts()


0    50000
Name: num_of_pictures, dtype: int64

In [12]:
autos['seller'].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [13]:
autos['offer_type'].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

In [15]:
autos.drop(columns=['offer_type', 'seller', 'num_of_pictures'], inplace=True )

In [16]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

    price  ab_test vehicle_type  registration_year    gearbox  power_PS  \
0  $5,000  control          bus               2004    manuell       158   
1  $8,500  control    limousine               1997  automatik       286   
2  $8,990     test    limousine               2009    manuell       102   
3  $4,350  control   kleinwagen               2007  automatik        71   
4  $1,350     test        kombi               2003    manuell         0   

    model   odometer  registration_month fuel_type       brand  \
0  andere  150,000km                   3       lpg     peugeot   
1     7er  150,000km                   6    benzin         bmw   
2    golf   70,000km                   7    benzin  volkswagen   
3  fortwo   70,000km                   6    benzin       smart   
4   focus  150,000km                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

In [17]:
autos['odometer'] = (autos['odometer']
                     .str.replace('km', '') 
                     .str.replace(',', '')
                     .astype(int)
                    )
autos.rename(columns = {'odometer': 'odometer_km'},inplace= True)
autos['odometer_km'].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

In [18]:
autos['price'] = (autos['price']
                  .str.replace('$', '')
                  .str.replace(',', '')
                  .astype(int)
                 )
autos['price'].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

### Exploring Odometer and Price

In [19]:
autos['odometer_km'].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

#### The values under odometer are rounded. And there are more high mileage than low ones

In [20]:
print(autos['price'].unique().shape)
print(autos['price'].describe())
autos['price'].value_counts().head(20)

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
700      395
4500     394
300      384
2200     382
950      379
Name: price, dtype: int64

#### The price values are also rounded.

In [21]:
autos['price'].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

In [22]:
autos['price'].value_counts().sort_index(ascending=True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64

We have observed off values under price. There can not be a car with value of 0 amount and having a value as high as 999999 is obvious that this is unrealistic. However it would make sence for a value to be 1 which is common with ebay autions. Consequently we would remove values between 1 and 350000

In [23]:
autos = autos[autos['price'].between(1,351000)]
autos['price'].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

In [24]:
autos.price.unique()

array([ 5000,  8500,  8990, ...,   385, 22200, 16995])

There are different ways of removing outlier, however we applied the quantile method by removing 1% of the lower band and 1% of the higher band of the prices
Removing the the outlier under the price column has lead to the reduction of observation from 50000 to 48072. Also the mean and the median are now reasonable. It has also reduce the minimum price to 1 whereas the maximum price is 35,890

In [25]:
autos[['date_crawled', 'ad_created', 'last_seen']]

date_crawled           ad_created            last_seen
0      2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1      2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2      2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3      2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4      2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50
5      2016-03-21 13:47:45  2016-03-21 00:00:00  2016-04-06 09:45:21
6      2016-03-20 17:55:21  2016-03-20 00:00:00  2016-03-23 02:48:59
7      2016-03-16 18:55:19  2016-03-16 00:00:00  2016-04-07 03:17:32
8      2016-03-22 16:51:34  2016-03-22 00:00:00  2016-03-26 18:18:10
9      2016-03-16 13:47:02  2016-03-16 00:00:00  2016-04-06 10:46:35
10     2016-03-15 01:41:36  2016-03-14 00:00:00  2016-04-07 03:16:21
11     2016-03-16 18:45:34  2016-03-16 00:00:00  2016-03-16 18:45:34
12     2016-03-31 19:48:22  2016-03-31 00:00:00  2016-04-06 14:17:52
13     2016-03-23 10:48:32  2016-03-23 00:00:00  2016-04-01 14:17:13
14     2016-03-23 11:50:46  2016-03-23 00:00:00  2016-04-01 15:46:47
15     2016-04-01 12:06:20  2016-04-01 00:00:00  2016-04-02 21:10:48
16     2016-03-16 14:59:02  2016-03-16 00:00:00  2016-03-18 05:29:37
17     2016-03-29 11:46:22  2016-03-29 00:00:00  2016-04-05 20:46:26
18     2016-03-26 19:57:44  2016-03-26 00:00:00  2016-04-02 12:16:41
19     2016-03-17 13:36:21  2016-03-17 00:00:00  2016-03-17 14:45:58
20     2016-03-05 19:57:31  2016-03-05 00:00:00  2016-03-07 05:46:46
21     2016-03-06 19:07:10  2016-03-06 00:00:00  2016-04-05 23:46:19
22     2016-03-28 20:50:54  2016-03-28 00:00:00  2016-04-01 06:45:30
23     2016-03-10 19:55:34  2016-03-10 00:00:00  2016-03-17 08:45:17
24     2016-04-03 11:57:02  2016-04-03 00:00:00  2016-04-07 13:16:50
25     2016-03-21 21:56:18  2016-03-21 00:00:00  2016-04-01 05:16:49
26     2016-04-03 22:46:28  2016-04-03 00:00:00  2016-04-05 23:46:48
28     2016-03-19 21:56:19  2016-03-19 00:00:00  2016-04-07 14:58:48
29     2016-04-02 12:45:44  2016-04-02 00:00:00  2016-04-02 12:45:44
30     2016-03-14 11:47:31  2016-03-14 00:00:00  2016-03-14 11:47:31
...                    ...                  ...                  ...
49968  2016-04-01 17:49:15  2016-04-01 00:00:00  2016-04-05 13:16:35
49969  2016-03-17 18:49:02  2016-03-17 00:00:00  2016-03-25 23:18:15
49970  2016-03-21 22:47:37  2016-03-21 00:00:00  2016-04-07 04:17:34
49971  2016-03-29 14:54:12  2016-03-29 00:00:00  2016-03-29 20:41:51
49972  2016-03-26 22:25:23  2016-03-26 00:00:00  2016-03-28 11:28:18
49973  2016-03-27 05:32:39  2016-03-27 00:00:00  2016-03-27 08:25:24
49975  2016-03-27 20:51:39  2016-03-27 00:00:00  2016-04-05 19:45:34
49976  2016-03-19 18:56:05  2016-03-19 00:00:00  2016-04-07 06:16:44
49977  2016-03-31 18:37:18  2016-03-31 00:00:00  2016-04-06 12:16:11
49978  2016-04-04 10:37:14  2016-04-04 00:00:00  2016-04-06 12:44:20
49979  2016-03-20 18:38:40  2016-03-20 00:00:00  2016-04-07 01:45:12
49980  2016-03-12 10:55:54  2016-03-12 00:00:00  2016-04-06 17:16:49
49981  2016-03-15 09:38:21  2016-03-15 00:00:00  2016-04-05 17:21:46
49982  2016-03-29 18:51:08  2016-03-29 00:00:00  2016-03-29 18:51:08
49983  2016-03-06 12:43:04  2016-03-06 00:00:00  2016-03-09 06:16:08
49985  2016-04-02 16:38:23  2016-04-02 00:00:00  2016-04-06 15:17:52
49986  2016-04-04 20:46:02  2016-04-04 00:00:00  2016-04-06 23:16:00
49987  2016-03-22 20:47:27  2016-03-22 00:00:00  2016-03-26 22:46:06
49988  2016-03-28 19:49:51  2016-03-28 00:00:00  2016-04-07 02:17:40
49989  2016-03-11 19:50:37  2016-03-11 00:00:00  2016-03-12 10:17:55
49990  2016-03-21 19:54:19  2016-03-21 00:00:00  2016-04-06 22:46:57
49991  2016-03-06 15:25:19  2016-03-06 00:00:00  2016-03-06 18:24:19
49992  2016-03-10 19:37:38  2016-03-10 00:00:00  2016-03-13 01:44:51
49993  2016-03-15 18:47:35  2016-03-15 00:00:00  2016-04-06 19:46:53
49994  2016-03-22 17:36:42  2016-03-22 00:00:00  2016-04-06 08:16:39
49995  2016-03-27 14:38:19  2016-03-27 00:00:00  2016-04-01 13:47:40
49996  2016-

In [26]:
(autos['date_crawled']
.str[:10]
.value_counts( normalize= True, ascending=True,dropna=False)
.sort_index()
)

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

The column is certified free from missing values and sorted in a descending order.

In [27]:
(autos['ad_created']
.str[:10]
.value_counts(normalize=True,ascending=True, dropna=False)
.sort_index()
)

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21 

The column is certified free from missing values and sorted in a descending order.

In [28]:
(autos['last_seen']
.str[:10]
.value_counts(normalize=True, dropna=False)
.sort_index()
)

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

The column is certified free from missing values and sorted in a descending order.

### Exploring and dealing with incorrect registratio year

Because a car can't be first registered before the listing was seen, any vehicle with a registration above 2016 is definately inaccurate.
we can eremove the listings with these values.
Let's see the percentage of these values

In [29]:
(~autos['registration_year'].between(1900,2016)).sum()/autos.shape[0]

0.038793369710697

Given that these values we about to remove are less than 4% of our data, we can remove these rows

In [30]:
autos = autos[autos['registration_year'].between(1900,2016)]
autos['registration_year'].value_counts(normalize=True).head(10)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
Name: registration_year, dtype: float64

In [31]:
autos.groupby('registration_year').min()

date_crawled  \
registration_year                        
1910               2016-03-11 22:37:01   
1927               2016-03-12 08:36:21   
1929               2016-03-09 16:51:17   
1931               2016-03-26 19:49:59   
1934               2016-03-15 14:57:07   
1937               2016-03-05 17:45:32   
1938               2016-03-16 12:37:38   
1939               2016-03-31 13:57:18   
1941               2016-03-09 10:54:11   
1943               2016-03-11 21:48:36   
1948               2016-03-08 20:50:10   
1950               2016-03-19 11:52:47   
1951               2016-03-15 14:47:26   
1952               2016-03-28 22:52:39   
1953               2016-03-19 13:51:49   
1954               2016-03-07 12:52:18   
1955               2016-03-12 21:54:53   
1956               2016-03-06 10:51:30   
1957               2016-03-12 23:57:10   
1958               2016-03-09 22:55:53   
1959               2016-03-05 14:45:08   
1960               2016-03-07 10:55:57   
1961               2016-03-07 12:55:27   
1962               2016-03-16 08:55:27   
1963               2016-03-09 18:37:55   
1964               2016-03-05 21:51:12   
1965               2016-03-09 21:49:13   
1966               2016-03-05 16:41:40   
1967               2016-03-06 23:52:19   
1968               2016-03-07 19:58:50   
...                                ...   
1987               2016-03-05 14:10:18   
1988               2016-03-05 18:52:55   
1989               2016-03-05 17:38:18   
1990               2016-03-05 15:42:33   
1991               2016-03-05 14:48:58   
1992               2016-03-05 14:28:08   
1993               2016-03-05 14:14:31   
1994               2016-03-05 14:12:36   
1995               2016-03-05 14:06:30   
1996               2016-03-05 14:07:45   
1997               2016-03-05 14:11:59   
1998               2016-03-05 14:08:00   
1999               2016-03-05 14:12:16   
2000               2016-03-05 14:08:05   
2001               2016-03-05 14:06:40   
2002               2016-03-05 14:09:58   
2003               2016-03-05 14:07:21   
2004               2016-03-05 14:07:26   
2005               2016-03-05 14:07:04   
2006               2016-03-05 14:20:00   
2007               2016-03-05 14:09:46   
2008               2016-03-05 14:09:58   
2009               2016-03-05 14:09:02   
2010               2016-03-05 14:09:05   
2011               2016-03-05 14:16:30   
2012               2016-03-05 14:21:29   
2013               2016-03-05 14:25:16   
2014               2016-03-05 14:40:08   
2015               2016-03-05 15:48:52   
2016               2016-03-05 14:21:30   

                                                                name  price  \
registration_year                                                             
1910                                                      Motorhaube      1   
1927                                         Essex_super_six__Ford_A  16500   
1929                                                      BMW_Andere  11500   
1931                               Ford_Model_A_Roadster_Deluxe_1931  27500   
1934                           Hanomag_rekord_15k_Suche_ersatz_teile   3000   
1937               Dodge_1937_Hotrod_Flathead_Scheunenfund_Oldtim...   4800   
1938                            Oldtimer_Sport___Cabrio_Georges_Irat   8500   
1939                                   Andere_DKW_F8_Cabrio_Limosine  26600   
1941                           Andere_GAZ_Oldtimergelaendewagen_1941   9500   
1943                                           Volkswagen__VW_Typ_82  41900   
1948                       Andere_Simca_5_Fourgonette_Kombilimousine  17500   
1950               Oldtimer_fuer_meine_Sammlung__Porsche_911_356_...      1   
1951                                              Land_Rover_Serie_I  25500   
1952                                                Oldtimer__IFA_F9   3800   
1953                                 Suche_Fachwissen_zu_Gutbrod_bus   6666   
1954                         Chevrol

In [32]:
autos.groupby('registration_year').max()

date_crawled  \
registration_year                        
1910               2016-04-04 00:36:17   
1927               2016-03-12 08:36:21   
1929               2016-03-09 16:51:17   
1931               2016-03-26 19:49:59   
1934               2016-03-19 22:51:25   
1937               2016-03-30 22:42:50   
1938               2016-03-16 12:37:38   
1939               2016-03-31 13:57:18   
1941               2016-03-20 17:51:49   
1943               2016-03-11 21:48:36   
1948               2016-03-08 20:50:10   
1950               2016-03-29 17:53:03   
1951               2016-03-20 20:48:25   
1952               2016-03-28 22:52:39   
1953               2016-03-19 13:51:49   
1954               2016-03-29 11:47:24   
1955               2016-04-01 09:47:52   
1956               2016-03-28 16:26:34   
1957               2016-03-20 15:54:00   
1958               2016-03-19 13:45:25   
1959               2016-04-02 01:57:41   
1960               2016-04-01 00:57:49   
1961               2016-04-04 19:52:35   
1962               2016-03-31 21:48:00   
1963               2016-04-04 21:38:51   
1964               2016-04-04 21:36:57   
1965               2016-04-04 10:37:42   
1966               2016-04-04 11:25:36   
1967               2016-04-04 20:48:18   
1968               2016-04-05 10:55:39   
...                                ...   
1987               2016-04-05 10:46:42   
1988               2016-04-05 09:51:04   
1989               2016-04-06 18:25:34   
1990               2016-04-07 09:36:25   
1991               2016-04-05 23:25:37   
1992               2016-04-07 03:36:25   
1993               2016-04-07 13:36:37   
1994               2016-04-06 12:36:34   
1995               2016-04-07 10:36:17   
1996               2016-04-07 13:36:20   
1997               2016-04-07 11:36:35   
1998               2016-04-07 13:06:18   
1999               2016-04-07 14:30:09   
2000               2016-04-07 03:36:33   
2001               2016-04-07 07:06:34   
2002               2016-04-07 14:36:44   
2003               2016-04-07 11:06:33   
2004               2016-04-07 12:06:23   
2005               2016-04-07 14:30:26   
2006               2016-04-07 10:36:36   
2007               2016-04-07 14:07:04   
2008               2016-04-07 14:36:55   
2009               2016-04-07 14:36:56   
2010               2016-04-07 12:06:19   
2011               2016-04-07 00:06:23   
2012               2016-04-07 01:36:16   
2013               2016-04-07 02:36:24   
2014               2016-04-06 11:36:16   
2015               2016-04-05 18:36:32   
2016               2016-04-07 11:36:24   

                                                                name   price  \
registration_year                                                              
1910                                                      Suche_Auto     599   
1927                                         Essex_super_six__Ford_A   16500   
1929                                                      BMW_Andere   11500   
1931                               Ford_Model_A_Roadster_Deluxe_1931   27500   
1934               Sehr_seltener_Oldtimer_Opel_1210_zum_Restaurieren    3350   
1937                                              Mercedes_Benz_170V   25500   
1938                            Oldtimer_Sport___Cabrio_Georges_Irat    8500   
1939                                   Andere_DKW_F8_Cabrio_Limosine   26600   
1941               Mercedes_Benz_L1500S_Wehrmacht_/_Luftwaffe___F...   26900   
1943                                           Volkswagen__VW_Typ_82   41900   
1948                       Andere_Simca_5_Fourgonette_Kombilimousine   17500   
1950                                        Seat_Leon_Spielzeug_Auto   10000   
1951               Mercedes_Benz_170_S_Cabriolet_A__Startklar_fue...  120000   
1952                                                Oldtimer__IFA_F9    3800   
1953                                 Suche_Fachwissen_zu_Gutbrod_bus    6666   
1954               

In [33]:
autos[(autos.registration_year>=1900) & (autos.registration_year<=2016)]

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
30     2016-03-14 11:47:31                         Peugeot_206_Unfallfahrzeug   
31     2016-03-14 16:53:09                          Noch_gut_erhaltenen_C_320   
...                    ...                                                ...   
49968  2016-04-01 17:49:15                  Mercedes_Benz_190_D_2.5_Automatik   
49969  2016-03-17 18:49:02             Nissan_X_Trail_2.2_dCi_4x4_Sport_m.AHZ   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine

In [34]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price  ab_test vehicle_type  registration_year    gearbox  power_PS  \
0   5000  control          bus               2004    manuell       158   
1   8500  control    limousine               1997  automatik       286   
2   8990     test    limousine               2009    manuell       102   
3   4350  control   kleinwagen               2007  automatik        71   
4   1350     test        kombi               2003    manuell         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere       150000                   3       lpg     peugeot   
1     7er       150000                   6    benzin         bmw   
2    golf        70000                   7    benzin  volkswagen   
3  fortwo        70000                   6    benzin       smart   
4   focus       150000                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

### Exploring Price By Brand

In [35]:
autos['brand'].value_counts(normalize=True)

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

These are the top 20 brands of use car that are being sold. Volkswagen apparently is at the top, which is followed closely by bmw. why is this so? could it be that these brands are old and the owners are desperate to do away with them for newer brands? This can be further investigated.

There are lots of brand that don't have a significant percentage of listing, so we will limit ouranalysis to brands representing more than 5% of total listing

In [36]:
brand_counts = autos['brand'].value_counts(normalize=True)
common_brands = brand_counts[brand_counts > .05].index
print(common_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [37]:
brand_mean_prices = {}


for brand in common_brands:
    brand_only = autos[autos['brand'] == brand]
    mean_price = brand_only['price'].mean()
    brand_mean_prices[brand] = int(mean_price)

brand_mean_prices




       

{'audi': 9336,
 'bmw': 8332,
 'ford': 3749,
 'mercedes_benz': 8628,
 'opel': 2975,
 'volkswagen': 5402}

Of the top 5 brands, there is a distict price gap:
* Audi, BMW and Mercedes Benz are more expensive
* Ford and Opel are less expensive
* Volkswagen is in between.

### Exploring Mileage

In [38]:
bmp_series = pd.Series(brand_mean_prices)
pd.DataFrame(bmp_series, columns=['mean_mileage'])

mean_mileage
audi                   9336
bmw                    8332
ford                   3749
mercedes_benz          8628
opel                   2975
volkswagen             5402

In [39]:
brand_mean_mileage = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_mileage = brand_only["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)

mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending=False)


In [40]:
brand_info = pd.DataFrame(mean_mileage,columns=['mean_mileage'])
brand_info

mean_mileage
bmw                  132572
mercedes_benz        130788
opel                 129310
audi                 129157
volkswagen           128707
ford                 124266

In [41]:
brand_info["mean_price"] = mean_prices
brand_info

mean_mileage  mean_price
bmw                  132572        8332
mercedes_benz        130788        8628
opel                 129310        2975
audi                 129157        9336
volkswagen           128707        5402
ford                 124266        3749

The more expensive cars have higher mileage whereas the less expensive cars have a lower mileage

In [47]:
# autos.loc[:, 'date_crawled'] \
        # .str.replace('-', '')